From 76299239ca928dd09eebe8d31dfec460478f5e09 Mon Sep 17 00:00:00 2001
From: Colin Jones <trptcolin@gmail.com>
Date: Wed, 11 Jan 2012 19:09:59 -0600
Subject: [PATCH] Allow custom buffer size in LineNumberingPushbackReader

---
 .../clojure/lang/LineNumberingPushbackReader.java  |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/src/jvm/clojure/lang/LineNumberingPushbackReader.java b/src/jvm/clojure/lang/LineNumberingPushbackReader.java
index ad81ce9..65c6d5f 100644
--- a/src/jvm/clojure/lang/LineNumberingPushbackReader.java
+++ b/src/jvm/clojure/lang/LineNumberingPushbackReader.java
@@ -32,6 +32,10 @@ public LineNumberingPushbackReader(Reader r){
 	super(new LineNumberReader(r));
 }
 
+public LineNumberingPushbackReader(Reader r, int size){
+	super(new LineNumberReader(r, size));
+}
+
 public int getLineNumber(){
 	return ((LineNumberReader) in).getLineNumber() + 1;
 }
--
1.7.5.4

